Add pagination block

swissspidy opened the pull request:
#1440 and #1460 inspired me to add a new block to support `<!--nextpage-->` in Gutenberg.

See #1328.

This PR does:

* add a new core/nextpage block
* update `special-comment-converter.js` to turn `<!--nexpage-->` into that new block
* add tests

Screenshots:

<img width="970" alt="screen shot 2017-06-26 at 21 46 53" src="https://user-images.githubusercontent.com/841956/27557497-1d9b0508-5aba-11e7-972a-5a1d9f665876.png">
<img width="337" alt="screen shot 2017-06-26 at 21 47 04" src="https://user-images.githubusercontent.com/841956/27557498-1dc066cc-5aba-11e7-9c8d-ce8fbd91bcc9.png">
<img width="267" alt="screen shot 2017-06-26 at 21 47 16" src="https://user-images.githubusercontent.com/841956/27557499-1dc290c8-5aba-11e7-9484-2919b4581915.png">

chrisgrabinski commented:
It might be useful to add page numbers to the block output. This would improve navigation inside the editor for long articles with multiple pages.

![page-numbers](https://user-images.githubusercontent.com/958819/27558475-6a01eb14-5ae7-11e7-8917-97c4827fd280.png)

dmsnell commented:
@swissspidy this looks fine to me, but yeah we may want to revisit. Open questions include:
 - how many of these exceptions are there where we have a magical HTML comment?
 - should we create a new top-level rule in the parser for the magic comments? (I started this way in #1460 but took it out as a "when we get there" kind of thing)
 - where should we attempt to track the stateful information, such as _which_ page number we're breaking? we can do it in the parser, though that's really something we want to avoid since it breaks the parsing algorithm. we can do it in the `parser.js` file, but we have to make sure we don't start proliferating hacks in there.

so far though this PR looks good to me. thanks!

youknowriad reviewed (commented):
Code looks good but for some reason the serializer is not working as expected (Try switching to the text mode you'll notice the `<!-- wp:core/nextpage /-->` comments instead of `<!--nextpage-->`

swissspidy commented:
Thanks @youknowriad, I fixed the serializer now and added a test to verify that. Needed to move some declarations in order to not get warnings b/c of missing attributes.

------

> how many of these exceptions are there where we have a magical HTML comment?

To my knowledge, there are only `<!--more(.*?)?-->`, `<!--noteaser-->` and `<!--nextpage-->`.

> should we create a new top-level rule in the parser for the magic comments? (I started this way in #1460 but took it out as a "when we get there" kind of thing)

Since there are only 3 special HTML comments, it might make sense to have 1 specific top-level rule for them in the parser.

> It might be useful to add page numbers to the block output. This would improve navigation inside the editor for long articles with multiple pages.

Good idea. I thought I saw some code in Gutenberg for such stuff the other day, but I can't find it right now.

youknowriad commented:
Looks like there's also an issue with the parser, try saving a post with this block and reloading the page.

swissspidy commented:
Ugh, indeed.

I just added a failing test in 35ef9f12a595168b14062a9377d064c163306803 to demonstrate this.

The test results in only 3 blocks, where the third block gets a raw content of `\n<p>Broccoli</p>\n<!--nextpage--><p>Romanesco</p>` instead of just `<p>Broccoli</p>` and the rest being split into more blocks.

swissspidy commented:
@youknowriad Do you perhaps have an idea regarding the parser and how I could fix it?

youknowriad commented:
@swissspidy I think we'd need to exclude the `WP_Block_NextPage` block from the `WP_Block_HTML` (line 51 of the `pegjs` file) (same as void block for example)

swissspidy commented:
Interesting, I haven't seen that before. I did that in 8ee3f50a575077bafbe1d8126e245cc9bca19025 but it doesn't seem to work. See https://travis-ci.org/WordPress/gutenberg/jobs/249624708#L2010-L2016

swissspidy commented:
Ugh, so the tests partly just failed because of #1460. I needed to add a `save()` method when registering the blocks in the test.

dmsnell commented on blocks/api/post.pegjs:
```diff
@@ -95,7 +112,7 @@ WP_Block_Balanced
 
 WP_Block_Html
-  = ts:(!WP_Block_Balanced !WP_Block_Void c:Any {
+  = ts:(!WP_Block_Balanced !WP_Block_Void !WP_Block_NextPage c:Any {
```
Should probably also put in `!WP_Tag_More` here until we fix up the root cause of this trouble

swissspidy replied:
Should I also rename `WP_Block_NextPage` to `WP_Tag_NextPage`?

dmsnell replied:
It's up to you. It would be nice if they were both the same. I chose `Tag` because of the exceptional nature of the tag, not knowing originally it would turn into a block. I think that difference is legitimate here since it's clearly not a block in the text.

Either way, could we normalize them both to match each other?

dmsnell commented on blocks/api/serializer.js:
```diff
@@ -111,6 +111,11 @@ export function serializeAttributes( attrs ) {
+	if ( 'core/nextpage' === blockName ) {
+		return '<!--nextpage-->';
+	}
+
```
would probably be nice to keep this next to the `more` tag. if it's up here to save on computation I would suggest that there's not much to save on this one special tag over the cost of serializing the document.

swissspidy replied:
I know it would be nice, but it's up there to not cause errors when calling `getSaveContent()` or  `getCommentAttributes()` because this block doesn't have any of that.

dmsnell replied:
is that problematic then that the block doesn't do it. maybe we've surfaced an oddity or bug?

it seems like other blocks might also have no attributes and if they break here we will want to make sure that doesn't happen.

swissspidy replied:
Hmm, I had another look and it seems like the problem just was the missing `save()` method in the test, no other warnings. So I guess that resolved itself.

nylen commented on blocks/api/test/parser.js:
```diff
@@ -282,5 +282,66 @@ describe( 'block parser', () => {
+			expect( parsed[ 0 ].attributes.content ).to.eql( '<p>Cauliflower</p>' );
+			expect( parsed[ 2 ].attributes.content ).to.eql( '<p>Broccoli</p>' );
+			expect( parsed[ 4 ].attributes.content ).to.eql( '<p>Romanesco</p>' );
+		} );
```
This last test case in particular is probably a better fit for a new file in `blocks/test/fixtures/`, for example `misc__multi-blocks-and-nextpage.html`.  This way it will be both easier to set up and more thoroughly tested.

swissspidy replied:
Sounds reasonable. Is there precedent for this or shall I just go on to create the first such `misc__…` file?

nylen replied:
This would be the first such fixture, yes.  There need to be many more before we are ready to release.

aduth commented on blocks/library/nextpage/index.js:
```diff
@@ -0,0 +1,41 @@
+/**
+ * WordPress dependencies
+ */
+import { __ } from 'i18n';
```
Note with #2172, these need to be updated to prefix the dependencies with `@wordpress/`. You will need to perform a rebase against the latest version of master and apply your changes:

```
git fetch origin
git rebase origin/master
```

gziolo commented:
@swissspidy, do you plan to continue the work started in this PR. From the comments looks like it was in pretty good shape a few months back. It might require some extra work to make it up to date.

mcsf commented:
@swissspidy, I'm interested in this per #5031. Let me know what the status is, or if you'd like to collaborate on it or reassign it. :)

swissspidy commented:
@mcsf Interesting! Would be happy to collaborate. I'll try to resolve conflicts this week and then we can iterate from there.

mcsf commented:
Hi, @swissspidy, how're things here? :)

swissspidy commented:
Works like a charm again. Demo:

<img width="1509" alt="screen shot 2018-02-20 at 13 26 52" src="https://user-images.githubusercontent.com/841956/36424545-ec5a9f94-1643-11e8-9f6f-123ccff2dbf6.png">
<img width="784" alt="screen shot 2018-02-20 at 13 27 49" src="https://user-images.githubusercontent.com/841956/36424546-ec7bd8da-1643-11e8-9521-6baeffd0c805.png">

Adding page numbers to the block was previously mentioned, but I think that could be added in a new PR if considered useful.

@mcsf Feel free to check it out now :)

mcsf commented:
@swissspidy, apologies for the delay.

Thanks for refreshing this branch. I wanted to have a couple of other PRs merged before getting back to you. The crux is #5061, which reworks the More block so as to not impact the parser. I'd like for the pagination block to follow the same pattern. In other words, it shouldn't add anything to the PEG parser, thus optimizing for low parser complexity, meaning better robustness, performance and separation of concerns.

A couple of comments to have in mind from that PR:
- https://github.com/WordPress/gutenberg/pull/5061#discussion_r169279282
- https://github.com/WordPress/gutenberg/pull/5061#issuecomment-367988651

If this change is course is frustrating, I understand. If you'd still like to take this on, I'd be quite grateful!

swissspidy commented:
As mentioned in that other PR, `special-comment-converter` might need to undergo some changes. Apart from that this should be fine for the moment.

@mcsf Mind having another look now?

mcsf commented on blocks/api/raw-handling/special-comment-converter.js:
```diff
@@ -19,10 +19,27 @@ const { COMMENT_NODE } = window.Node;
+		return;
+	}
+
+	if ( node.nodeValue.indexOf( 'nextpage' ) === 0 ) {
```
I don't believe `nextpage` accepts any "arguments" within the comment, so, in order to match core, this should be a full string eq check: `node.nodeValue === 'nextpage'`.

mcsf commented on blocks/api/raw-handling/special-comment-converter.js:
```diff
@@ -19,10 +19,27 @@ const { COMMENT_NODE } = window.Node;
+		parent.appendChild( nextPage );
+		node.parentNode.removeChild( node );
+
+		return;
```
Instead of repeating the parent finding and the node operations for both comment types, we could refactor this. The simplest way I can think of may be:

```js
let newNode;

if ( foundNextpage ) {
  newNode = createNextpage();
} else if ( foundMore ) {
  newNode = createMore( ...args );
}

if ( newNode ) {
  let parent = …; while ( … ) parent = parent.Node;
  parent.appendChild( … ); node.parentNode.removeChild( … );
  return;
}
```

or, actually better:

```js
if ( foundNextpage ) {
  appendToBody( node, createNextpage() );
  remove( node );
  return;
}
if ( foundMore ) {
  appendToBody( node, createMore( … ) );
  remove( node );
}

function appendToBody( processedNode, newNode ) {
  // Append it to the top level for later conversion to blocks
  let parent = node.parentNode;
  while ( parent.nodeName !== 'BODY' ) {
    parent = parent.parentNode;
  }

  parent.appendChild( nextPage );
}

function remove( node ) {
  node.parentNode.removeChild( node );
}
```

mcsf commented on blocks/library/nextpage/index.js:
```diff
@@ -0,0 +1,59 @@
+	title: __( 'Page break' ),
+
+	description: __( 'This block allows you to set break points on your post. Visitors of your blog are then presented with content split into multiple pages.' ),
+
```
A `keyword` property here would be useful to include things like `next page`, helping with discovery in the block inserters.

mcsf commented on blocks/library/nextpage/index.js:
```diff
@@ -0,0 +1,59 @@
+	save() {
+		return (
+			<RawHTML>
+				{ [ '<!--nextpage-->' ].join( '\n' ) }
```
No need for the array + join operation :)

```js
<RawHTML>
  { '<!--nextpage-->' }
</RawHTML>
```

mcsf reviewed (commented):
Thanks for the super quick response! This is looking good. I've left some comments.

swissspidy commented:
@mcsf I've incorporated your feedback now.

At the same time I realised that appending these elements to the body totally breaks the posts. This worked way better before #5061 and is now really frustrating for users. Put simply, I'd say this is not acceptable. Especially when it can be done better. Replacing the existing nodes makes much more sense, although still doesn't catch things like `<p>Foo<!--nextpage-->Bar</p>`.

Here are some examples:

Page breaks before conversion:
<img width="728" alt="page before" src="https://user-images.githubusercontent.com/841956/36901456-96aa428e-1e27-11e8-9943-ab19e69a3e96.png">

Page breaks after conversion:
<img width="734" alt="page after" src="https://user-images.githubusercontent.com/841956/36901466-9e98dde8-1e27-11e8-97be-4a9b0afdda51.png">

More tag before conversion:
<img width="687" alt="more before" src="https://user-images.githubusercontent.com/841956/36901477-a740a20a-1e27-11e8-9326-cbd6284a57e5.png">

More tag after conversion:
<img width="759" alt="more after" src="https://user-images.githubusercontent.com/841956/36901473-a45e4dd0-1e27-11e8-85c8-ad169e9fe8e3.png">

swissspidy commented:
Here's more or less what I had in mind for improving the node replacement:

```diff
diff --git blocks/api/raw-handling/special-comment-converter.js blocks/api/raw-handling/special-comment-converter.js
index ac2e1bc7..85fa5aad 100644
--- blocks/api/raw-handling/special-comment-converter.js
+++ blocks/api/raw-handling/special-comment-converter.js
@@ -24,8 +24,7 @@ export default function( node ) {
 	}
 
 	if ( node.nodeValue === 'nextpage' ) {
-		appendToBody( node, createNextpage() );
-		remove( node );
+		replace( node, createNextpage() );
 		return;
 	}
 
@@ -51,8 +50,7 @@ export default function( node ) {
 			}
 		}
 
-		appendToBody( node, createMore( customText, noTeaser ) );
-		remove( node );
+		replace( node, createMore( customText, noTeaser ) );
 	}
 }
 
@@ -76,14 +74,13 @@ function createNextpage() {
 	return node;
 }
 
-function appendToBody( processedNode, newNode ) {
-	// Append it to the top level for later conversion to blocks
-	let parent = processedNode.parentNode;
-	while ( parent.nodeName !== 'BODY' ) {
-		parent = parent.parentNode;
-	}
+function replace( processedNode, newNode ) {
+	insertAfter( newNode, processedNode.parentNode );
+	remove( processedNode );
+}
 
-	parent.appendChild( newNode );
+function insertAfter( newNode, referenceNode ) {
+	referenceNode.parentNode.insertBefore( newNode, referenceNode.nextSibling );
 }
 
 function remove( node ) {
```

mcsf commented:
Hey, @swissspidy, apologies for the delay, and great work here.

> At the same time I realised that appending these elements to the body totally breaks the posts. This worked way better before #5061 and is now really frustrating for users. Put simply, I'd say this is not acceptable. Especially when it can be done better. […]

Totally agree. I've tested your patch and it works great. I also wrote some tests for it. If you don't mind, I'll push a PR — so that we can fix More quickly — and give you props for it. I'll circle back for a review of Next Page.

swissspidy commented:
Rebased against master after #5538.

@mcsf Mind having another look?

mcsf commented:
Hey, I’ll have a look after we release the next version, so tomorrow
or Thursday. :-)

mcsf commented on blocks/library/nextpage/editor.scss:
```diff
@@ -0,0 +1,40 @@
+.wp-block-nextpage > span:after {
+	left: 100%;
+	margin-left: 20px;
+}
```
This file is a lot different from `blocks/library/more/editor.scss`. I prefer the final result for Next Page, and we'd probably want to converge a bit on the two stylesheets. Any advice from @jasmussen is much appreciated, as always.

mcsf reviewed (commented):
This looks great, thanks for stewarding and for your patience! Left some notes.

jasmussen commented on blocks/library/nextpage/index.js:
```diff
@@ -0,0 +1,61 @@
+
+	description: __( 'This block allows you to set break points on your post. Visitors of your blog are then presented with content split into multiple pages.' ),
+
+	icon: 'admin-page',
```
Yep, that icon is solid.

mcsf commented:
> I think perhaps it'd be good to merge this in as is, and then look at refining/unifying the styles separately. I can take that on as a quick smallish PR.

@jasmussen, that'd be awesome!

@swissspidy, seems like we're good to go, then. One last thing: can you rebase this entire branch into one-to-three neat commits against the latest `master`? Once you've got it, I'll merge it. 🙇

swissspidy committed:
Add new pagination block

swissspidy commented:
@mcsf Should look better now with just one commit.

mcsf commented:
This is great! Thanks for your perseverance, @swissspidy. :)
